cpp/src/community/detail/common_methods.cuh

```diff
@@ -0,0 +1,370 @@
+  using weight_t = typename graph_view_t::weight_type;
+
+  weight_t sum_degree_squared = thrust::transform_reduce(
+    handle.get_thrust_policy(),
+    cluster_weights.begin(),
+    cluster_weights.end(),
+    [] __device__(weight_t p) { return p * p; },
+    weight_t{0},
+    thrust::plus<weight_t>());
+
+  if constexpr (graph_view_t::is_multi_gpu) {
+    sum_degree_squared = host_scalar_allreduce(
+      handle.get_comms(), sum_degree_squared, raft::comms::op_t::SUM, handle.get_stream());
+  }
+
+  weight_t sum_internal = transform_reduce_e(
+    handle,
+    graph_view,
+    graph_view_t::is_multi_gpu
+      ? src_clusters_cache.view()
+      : detail::edge_major_property_view_t<vertex_t, vertex_t const*>(next_clusters.begin()),
+    graph_view_t::is_multi_gpu ? dst_clusters_cache.view()
+                               : detail::edge_minor_property_view_t<vertex_t, vertex_t const*>(
+                                   next_clusters.begin(), vertex_t{0}),
+    [] __device__(auto, auto, weight_t wt, auto src_cluster, auto nbr_cluster) {
+      if (src_cluster == nbr_cluster) {
+        return wt;
+      } else {
+        return weight_t{0};
+      }
+    },
+    weight_t{0});
+
+  weight_t Q = sum_internal / total_edge_weight -
+               (resolution * sum_degree_squared) / (total_edge_weight * total_edge_weight);
+
+  return Q;
+}
+
+template <typename vertex_t, typename edge_t, typename weight_t, bool multi_gpu>
+cugraph::graph_t<vertex_t, edge_t, weight_t, false, multi_gpu> graph_contraction(
+  raft::handle_t const& handle,
+  cugraph::graph_view_t<vertex_t, edge_t, weight_t, false, multi_gpu> const& graph_view,
+  raft::device_span<vertex_t> labels)
+{
+  auto [new_graph, numbering_map] = coarsen_graph(handle, graph_view, labels.data());
+
+  auto new_graph_view = new_graph.view();
+
+  rmm::device_uvector<vertex_t> numbering_indices(numbering_map.size(), handle.get_stream());
+  detail::sequence_fill(handle.get_stream(),
+                        numbering_indices.data(),
+                        numbering_indices.size(),
+                        new_graph_view.local_vertex_partition_range_first());
+
+  relabel<vertex_t, multi_gpu>(
+    handle,
+    std::make_tuple(static_cast<vertex_t const*>(numbering_map.begin()),
+                    static_cast<vertex_t const*>(numbering_indices.begin())),
+    new_graph_view.local_vertex_partition_range_size(),
+    labels.data(),
+    labels.size(),
+    false);
+
+  return std::move(new_graph);
+}
+
+template <typename graph_view_t>
+void update_by_delta_modularity(
+  raft::handle_t const& handle,
+  graph_view_t const& graph_view,
+  typename graph_view_t::weight_type total_edge_weight,
+  typename graph_view_t::weight_type resolution,
+  rmm::device_uvector<typename graph_view_t::weight_type>& vertex_weights_v,
+  rmm::device_uvector<typename graph_view_t::vertex_type>& cluster_keys_v,
+  rmm::device_uvector<typename graph_view_t::weight_type>& cluster_weights_v,
```

Comment:
So, this function updates clustering by delta modularity AND based on the updated clustering, updates per cluster weight sums as well. May better carve out the per cluster weight sum update part from this function?

If we remove
```
  std::tie(cluster_keys_v, cluster_weights_v) = cugraph::transform_reduce_e_by_src_key(
    handle,
    graph_view,
    edge_src_dummy_property_t{}.view(),
    edge_dst_dummy_property_t{}.view(),
    graph_view_t::is_multi_gpu
      ? src_clusters_cache.view()
      : detail::edge_major_property_view_t<vertex_t, vertex_t const*>(next_clusters_v.data()),
    detail::return_edge_weight_t<vertex_t, weight_t>{},
    weight_t{0});
```
`cluster_keys_v` and `cluster_weights_v` can be const as well, and it might be much easier to understand the input and output.

And if we further carve out the cache update parts, this function can take `next_clusters_v` as an R-value, and return the updated `next_clusters_v`. Then, all the input parameters will become const reference or scalars. This might be more intuitive.

Reply:
I agree with @seunghwak. It would make it easier to follow then. Also in that case it would be nicer to change the function name as well.

Reply:
Yes, I think I can refactor that as well.  That's probably an artifact of the implementation prior to using the primitives.  IIRC, I had an optimization in the original SG implementation that allowed me to skip a few of these steps if I did it all together.  Clearly now they can be separated.

Reply:
Refactored in next push